designs/capacity-reservations.md

```diff
@@ -0,0 +1,143 @@
+
+Attempting to launch this would result in a success for a single instance and an failure for the other 9,999. The next scheduling loop would remediate this, but this results in a lot of extra, unneeded work.
+
+A better way to model this would be to track the available instance count as a numerical value associated with an instance type offering. In this modeling, the scheduler could count the number of simulated NodeClaims that might use the offering and know that it can't simulate NodeClaims into particular offerings once they hit their cap.
+
+Today, we already have an [`available` field](https://github.com/kubernetes-sigs/karpenter/blob/bcd33e924905588b1bdecd5413dc7b268370ec4c/pkg/cloudprovider/types.go#L236) attached to instance type offerings. This field is binary and only tells us whether the instance is or isn't available. With the introduction of `karpenter.sh/capacity-type: reserved` offerings, we could extend this field to be an integer rather than a boolean. This would allow us to exactly represent the number of available instances that can be launched into the offering. Existing spot and on-demand offerings would model this `available` field as `MAX_INT` for current `true` values and `0` for `false` values.
+
+An updated version of the instance type offerings would look like:
+
+```yaml
+name: c5.large
+offerings:
+  - price: 0.085
+    available: 5
+    requirements:
+      ...
+      - key: karpenter.sh/capacity-type
+        operator: In
+        values: ["reserved"]
+  - price: 0.085
+    available: 4294967295
+    requirements:
+      ...
+      - key: karpenter.sh/capacity-type
+        operator: In
+        values: ["on-demand"]
+  - price: 0.0315
+    available: 4294967295
+    requirements:
+      ...
+      - key: karpenter.sh/capacity-type
+        operator: In
+        values: ["spot"]
+```
+
+### Consolidation
+
+_Note: Some excerpts taken from [`aws/karepnter-provider-aws` RFC](https://github.com/aws/karpenter-provider-aws/blob/main/designs/odcr.md#consolidation)._
+
+#### Consolidating into Capacity Reserved Instances
+
+Karpenter would need to update its consolidation algorithm to ensure that consolidating between a `spot` and/or `on-demand` capacity type to a reserved capacity type is always preferred. This can be done during the cost-checking step. When evaluating cost-savings, if we are able to consolidate all existing nodes into a `reserved` capacity type node, we will choose to do so.
+
+#### Consolidating between Capacity Reservations
```

Comment:
Should we include this detail about how we include price for ordering, but still prioritize reserved instances for scheduling?

https://github.com/aws/karpenter-provider-aws/blob/8e7eff5cc6318274a56bd2987db3b633af0b3539/designs/odcr.md?plain=1#L371

We didn't iron out in that ODCR if this would be handled by the cloudprovider or the core library. While we could reduce the price in the cloudprovider when constructing the offering, that requires each cloudprovider to be aware of the internal requirements of the scheduler and disruption controller.

Reply:
Yeah, as I was thinking through it and writing this round of the RFC, I realized that CloudProviders shouldn't have to do anything fancy with the price. We could just know to prioritize the capacity type in the upstream code. If this sounds reasonable to you, that's what I'm thinking we should go with here. We can make changes to the downstream RFC if we want as well to match that.

Reply:
Sounds good, wasn't sure if the omission of that detail here implied it was going to be handled downstream. We can probably omit the technique for maintaining ordering as an implementation detail.

Reply:
Agreed.